setup.py

```diff
@@ -853,202 +766,32 @@ def run(self):
-    main_link_args.append(os.path.join(lib_path, 'libtorch.dylib'))
+    main_link_args.append(os.path.join(lib_path, 'libtorch_python.dylib'))
 else:
-    main_link_args.append(os.path.join(lib_path, 'libtorch.so'))
-main_sources = [
-    "torch/csrc/DataLoader.cpp",
-    "torch/csrc/Device.cpp",
-    "torch/csrc/Dtype.cpp",
-    "torch/csrc/DynamicTypes.cpp",
-    "torch/csrc/Exceptions.cpp",
-    "torch/csrc/TypeInfo.cpp",
-    "torch/csrc/Generator.cpp",
-    "torch/csrc/Layout.cpp",
-    "torch/csrc/Module.cpp",
-    "torch/csrc/PtrWrapper.cpp",
-    "torch/csrc/Size.cpp",
-    "torch/csrc/Storage.cpp",
-    "torch/csrc/autograd/functions/init.cpp",
-    "torch/csrc/autograd/generated/python_functions.cpp",
-    "torch/csrc/autograd/generated/python_nn_functions.cpp",
-    "torch/csrc/autograd/generated/python_torch_functions.cpp",
-    "torch/csrc/autograd/generated/python_variable_methods.cpp",
-    "torch/csrc/autograd/init.cpp",
-    "torch/csrc/autograd/python_anomaly_mode.cpp",
-    "torch/csrc/autograd/python_cpp_function.cpp",
-    "torch/csrc/autograd/python_engine.cpp",
-    "torch/csrc/autograd/python_function.cpp",
-    "torch/csrc/autograd/python_hook.cpp",
-    "torch/csrc/autograd/python_legacy_variable.cpp",
-    "torch/csrc/autograd/python_variable.cpp",
-    "torch/csrc/autograd/python_variable_indexing.cpp",
-    "torch/csrc/byte_order.cpp",
-    "torch/csrc/jit/batched/BatchTensor.cpp",
-    "torch/csrc/jit/init.cpp",
-    "torch/csrc/jit/passes/onnx.cpp",
-    "torch/csrc/jit/passes/onnx/fixup_onnx_loop.cpp",
-    "torch/csrc/jit/passes/onnx/prepare_division_for_onnx.cpp",
-    "torch/csrc/jit/passes/onnx/peephole.cpp",
-    "torch/csrc/jit/passes/to_batch.cpp",
-    "torch/csrc/jit/python_arg_flatten.cpp",
-    "torch/csrc/jit/python_interpreter.cpp",
-    "torch/csrc/jit/python_ir.cpp",
-    "torch/csrc/jit/python_tracer.cpp",
-    "torch/csrc/jit/script/init.cpp",
-    "torch/csrc/jit/script/lexer.cpp",
-    "torch/csrc/jit/script/module.cpp",
-    "torch/csrc/jit/script/python_tree_views.cpp",
-    "torch/csrc/nn/THNN.cpp",
-    "torch/csrc/onnx/init.cpp",
-    "torch/csrc/serialization.cpp",
-    "torch/csrc/tensor/python_tensor.cpp",
-    "torch/csrc/utils.cpp",
-    "torch/csrc/utils/cuda_lazy_init.cpp",
-    "torch/csrc/utils/invalid_arguments.cpp",
-    "torch/csrc/utils/object_ptr.cpp",
-    "torch/csrc/utils/python_arg_parser.cpp",
-    "torch/csrc/utils/tensor_apply.cpp",
-    "torch/csrc/utils/tensor_conversion_dispatch.cpp",
-    "torch/csrc/utils/tensor_dtypes.cpp",
-    "torch/csrc/utils/tensor_flatten.cpp",
-    "torch/csrc/utils/tensor_layouts.cpp",
-    "torch/csrc/utils/tensor_list.cpp",
-    "torch/csrc/utils/tensor_new.cpp",
-    "torch/csrc/utils/tensor_numpy.cpp",
-    "torch/csrc/utils/tensor_types.cpp",
-    "torch/csrc/utils/tuple_parser.cpp",
-]
+    main_link_args.append(os.path.join(lib_path, 'libtorch_python.so'))
+main_sources = ["torch/csrc/stub.cpp"]
+main_link_args.extend(CAFFE2_LIBS)
```

Comment:
before this diff, _C.so and libcaffe2.so define some of the same symbols, and it's important that _C.so is loaded before libcaffe2.so so that the versions in _C.so get used. This automatically happens because we load _C.so directly, and libcaffe2.so is brought in as a dependency (though I suspect it might be possible to break by importing caffe2 first in the same process).

after this diff, libtorch_python.so and libcaffe2.so define some of the same symbols. We directly load the _C.so stub, which brings both of these in as dependencies. We have to make sure that symbols continue to be looked up in libtorch_python.so first, by making sure it comes before libcaffe2.so in the linker command.

Reply:
Can this live as an actual comment in the source code? It will be very hard to find the PR comment in the future.

Reply:
"this diff" isn't as well defined in a source file, but I can probably do something about it